### ISSUE TITLE
[FNA] [HARD] Replace MGFX with MojoShader#

### ISSUE THREAD

**Issue by flibitijibibo** [flibitijibibo/FNA#247]
Difficulty: This task is considered a HARD task. If MEDIUM tasks separate the men from the boys, HARD tasks separate Dark Souls fans from actual serial killers that wear proper armor for reasons yet to be explained.

Right now we're using MonoGame's MGFX format for shader support in FNA. This is problematic for a number of reasons:

- MGFX does not support the original XNB shaders!
- MGFX shaders require the 2MGFX tool to build, meaning we have to maintain a content builder.
- MGFX shaders for OpenGL aren't entirely great; there are lots of minor issues with how the Effect API tries to interop with the GLSL output we get from 2MGFX's custom MojoShader fork (which is very very old).

MGFX has its purpose for MonoGame, but for an XNA4 preservation project, it's really important that we support the original game data as much as possible, and that means we need to support the Effect files in their original form.

This is nice in that FNA ports no longer need to rebuild, but it's also nice that we don't have to maintain a content building tool as well... you can build effects with either the XNA tools or Microsoft's standalone Effect compiler, fxc.exe.

Something Ryan wrote a few years back for his Shank Linux/Mac ports was initial support for Effect files within MojoShader itself. While it's not complete, it's definitely a good base for what we need to complete this task. I've been working on this in a sandbox over here:

https://github.com/flibitijibibo/EffectRE

I'm basically just taking XNB shaders from the Windows versions of my ports, and taking them apart with UnXNB and MEPTool to ensure that we support the compiled Effect. This is a good place to start if you want to work on getting the file parsing complete.

The next step is to write an API to work with the parsed Effects. The actual shader support should be pretty simple, since that's what MojoShader has always done, but we need to write the entry points to nicely interop with the Effect data in a way that matches the original Effect Framework API. Again, all of this is in MojoShader itself... FNA's not involved quite yet. We're basically finishing Ryan's work and making a solid API for other projects to use, whether they're XNA-based or not.

Where FNA _will_ get involved is when we take out MGFX and replace it with MojoShader interop, using MojoShader#:

https://github.com/flibitijibibo/MojoShader-CS

This is based on the current header file, but as we finish the Effect API, we'll need to update MojoShader# accordingly. If we do our job correctly, most of the FNA-side of things will simply be hooking up the C# API to MojoShader entry points, with minimal fuss overall. Then the GL-specific stuff will simply go into the OpenGLDevice, like all the other GL calls in FNA.

So, here's the task list:
- Complete MOJOSHADER_parseEffect() function
- Write Effect API for MojoShader
- Examine/finish preshader support within MojoShader
- Submit patches to Ryan
- Update MojoShader# for above work
- Replace MGFX with MojoShader#, rewriting Graphics/Effect/ and removing Graphics/Shader/

I also need to get Ryan to merge another patch from some work I did a couple years ago ( http://flibitijibibo.com/mojoshader_diff.txt ), but I'll deal with that separately.

CC: @rcgordon, in case Ryan and FNA contributors need to sync up on anything here.

**Comment by flibitijibibo** [comment]
Just updated MojoShader and MojoShader_Win32 for the first time in a while... this marks Apotheon as working with the MojoShader# rewrite as well. Codename OhGodNo is still too messy for me to tell where the real bugs are, but we're definitely getting close to this getting merged. I've decided to merge when either of those two ports release, since I don't know what the order's going to be anymore.

**Comment by flibitijibibo** [comment]
Since I'm hosting these for Apotheon anyway, here are some Linux/Mac binaries too:

http://www.flibitijibibo.com/mojoshader32.tar.bz2
http://www.flibitijibibo.com/mojoshader64.tar.bz2
http://www.flibitijibibo.com/mojoshaderOSX.tar.bz2

Now it should be easier for everyone to test `mojoshader-cs`, which should be getting merged in the next few weeks.

**Comment by flibitijibibo** [comment]
This will be merged into `master` when Apotheon Linux/Mac ships:

https://plus.google.com/+flibitijibibo/posts/FWyMi7Ex14u

**Comment by flibitijibibo** [comment]
Apotheon Linux/Mac is now live, so this is now in mainline:

https://github.com/flibitijibibo/FNA/commit/8bc84bfc51436f89a99ce77c62bacb68d236069d

(Don't open this link unless you're nostalgic about Crysis on the machine you owned in 2007)

**Comment by Aranda** [comment]
>  I've provided a precompiled binary here for Windows developers to test with: http://flibitijibibo.com/MojoShader_Win32.zip

Have you still got some MojoShader binaries hosted? I'm testing out Square Heroes with the very latest FNA master and it seems I need them. Also wondering if this should Just Work (tm) with existing effects built with 2MGFX?

**Comment by flibitijibibo** [comment]
With the Effect rewrite now in, the binaries are now in the usual fnalibs archive:

http://www.flibitijibibo.com/fnalibs.tar.bz2

**Comment by Aranda** [comment]
Thanks Ethan.

BTW, not sure if this is a bug or expected, but it seems the latest MonoGame content changes don't support legacy Xml assets compiled with Xna's auto-Xnb-serialization. That's my guess anyway as I'm getting an incorrect ContentTypeReader index from Read7BitEncondedInt() in ListReader<T>.Read().

**Comment by flibitijibibo** [comment]
@Aranda If you can identify the revision where it broke I can check it out. Odds are it's a MonoGame-specific thing, since I've not had to touch the Content namespace in quite a few months. List history, for reference:

https://github.com/flibitijibibo/FNA/commits/master/src/Content/ContentReaders/ListReader.cs

If it breaks and it's XNA4 content, the MonoGame team will want to know.

**Comment by Aranda** [comment]
Ok scratch that last comment. I think my content file is just out of date :|

**Comment by flibitijibibo** [comment]
Hehe, it happens.

If it's built with the MonoGame pipeline, you may have to update on their side as well. @tgjones in particular has been doing some mighty fine work both on the pipeline side and on general XNA accuracy things, so if your stuff is built with their tools, it'd be worth trying with XNA4 as well to see if they're on the right track.

**Comment by Aranda** [comment]
Yeah I've been following the pipeline tool's progress with much interest, but haven't made the switch yet (still living in dummy XNA4 game plus bat-file copy land).

Incidentally, how do I go about compiling effects for your MojoShader change?

**Comment by flibitijibibo** [comment]
If you're building with XNA already, no change is necessary. Just use those XNB files and you're all set. When outside of the pipeline, you can build with fxc.exe, found in the DirectX SDK.
